crates/shared/src/current_block/eth_call.rs

```diff
@@ -0,0 +1,71 @@
+use super::{BlockInfo, BlockNumberHash, BlockRetrieving, RangeInclusive};
+use crate::ethrpc::Web3;
+use anyhow::{ensure, Context as _, Result};
+use contracts::support::FetchBlock;
+use web3::types::{BlockNumber, CallRequest, H256, U256};
+
+/// An `eth_call`-based block fetcher.
+///
+/// This can be used for nodes where `eth_getBlockBy*` and `eth_blockNumber`
+/// calls return the latest block for which a header is available even if the
+/// state isn't. This can be an issue for our services where we will try to
+/// update internal state to the new block even if the node we are connected to
+/// does not have the block available.
+pub struct BlockRetriever(pub Web3);
+
+#[async_trait::async_trait]
+impl BlockRetrieving for BlockRetriever {
+    async fn current_block(&self) -> Result<BlockInfo> {
+        let return_data = self
```

Comment:
This just moved the code introduced in #821 from `current_block.rs` to a separate type specific for `eth_call`-based block retrieving.